src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -0,0 +1,240 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+#if ES_BUILD_PCL
+    using System.Threading.Tasks;
+#endif
+
+#if ES_BUILD_STANDALONE
+namespace Microsoft.Diagnostics.Tracing
+#else
+namespace System.Diagnostics.Tracing
+#endif
+{
+    internal class CounterGroup
+    {
+        private readonly EventSource _eventSource;
+        private readonly List<BaseCounter> _counters;
+
+        internal CounterGroup(EventSource eventSource)
+        {
+            _eventSource = eventSource;
+            _counters = new List<BaseCounter>();
+            RegisterCommandCallback();
+        }
+
+        internal void Add(BaseCounter eventCounter)
+        {
+            lock (this) // Lock the CounterGroup
+                _counters.Add(eventCounter);
+        }
+
+        internal void Remove(BaseCounter eventCounter)
+        {
+            lock (this) // Lock the CounterGroup
+                _counters.Remove(eventCounter);
+        }
+
+        #region EventSource Command Processing
+
+        private void RegisterCommandCallback()
+        {
+            _eventSource.EventCommandExecuted += OnEventSourceCommand;
+        }
+
+        private void OnEventSourceCommand(object sender, EventCommandEventArgs e)
+        {
+            if (e.Command == EventCommand.Enable || e.Command == EventCommand.Update)
+            {
+                string valueStr;
+                float value;
+                if (e.Arguments.TryGetValue("EventCounterIntervalSec", out valueStr) && float.TryParse(valueStr, out value))
+                {
+                    // Recursion through EventSource callbacks possible.  When we enable the timer
```

Comment:
Should we open an issue about this?

Reply:
Yeah I think we should. I would not expect EventListener callbacks to ever be synchronous/recursive with the event that triggered them. If that is occuring it sounds like a recipe for deadlocks that is not specific to this code. For example this shouldn't deadlock and if EventListener is synchronous it might:

    object m_lock = new object();
    lock(m_lock) { EventSource.Write("Something"); }
    override EventListener.OnEventWritten(...) { lock(m_lock) { Console.WriteLine(event); } }

Reply:
Opened https://github.com/dotnet/coreclr/issues/23126